### PR TITLE
Mnaveau/solo calibration

### PR DESCRIPTION
**Pull request by MaximilienNaveau**
## What Changed?

- Fix the demo_testsstand_calibration adding the parsing of the robot_properties config file
- Fix the demo_solo_calibration adding the parsing of the robot_preperties_solo config file
- Added calibration procedure that helps doing the mechanical procedure for the Teststand and Solo.

## How did I tested?

- I ran the calibration procedures on both robots.

### CODE REVIEW

File: demos/demo_solo_calibration.cpp [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+static THREAD_FUNCTION_RETURN_TYPE control_loop(void* robot_void_ptr)
+{
+    Solo& robot = *(static_cast<Solo*>(robot_void_ptr));
+
+    Eigen::Vector2d joint_index_to_zero;
+    joint_index_to_zero[0] = 0.0;
+    joint_index_to_zero[1] = 0.0;
+    joint_index_to_zero[2] = 0.0;
+    joint_index_to_zero[3] = 0.0;
```
[the remark this answers is not in the document]
Comment (mwuethri) [reply demos/demo_solo_calibration.cpp]: also, from the variable name it is not clear to me what is the meaning of this

File: demos/demo_solo_calibration.cpp [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+    rt_printf("controller is set up \n");
+
+    thread.create_realtime_thread(&control_loop, &robot);
+
+    rt_printf("control loop started \n");
+
+    while (true)
+    {
+        real_time_tools::Timer::sleep_sec(0.01);
+    }
```
Comment (mwuethri) [inline demos/demo_solo_calibration.cpp]: why is this while loop here? shouldn't you rather join the thread? or if there is a while loop, shouldn't it be while(!StopDemos)?

File: include/blmc_robots/blmc_joint_module.hpp [hunk 3]
```diff
@@ -648,6 +648,12 @@ class BlmcJointModules
                          
         } while(current_time < (final_time + sampling_period));
 
+        for(unsigned i = 0; i < COUNT; i++)
+        {
+            modules_[i]->set_torque(0.0);
+            modules_[i]->send_torque();
+        }
```
Comment (mwuethri) [inline include/blmc_robots/blmc_joint_module.hpp]: why was this block added? i think this should probably be in a separate pull request, and ideally Felix should check that since i am not very clear about how this function works.

File: include/blmc_robots/solo.hpp [hunk 4]
```diff
@@ -63,104 +56,73 @@ class Quadruped
...
-   * @brief get_motor_encoder_indexes
-   * @return the position of the index of the encoders a the motor level
-   * WARNING !!!!
-   * The method <acquire_sensors>"()" has to be called
-   * prior to any getter to have up to date data.
+   * @brief get_start2hardstop_offsets_
+   * @return the position where the robot should be in "zero" configuration
    */
-  const Eigen::Ref<Vector8d> get_motor_encoder_indexes()
+  const Eigen::Ref<Vector8d> get_start2hardstop_offsets_()
```
Comment (mwuethri) [inline include/blmc_robots/solo.hpp]: why is there an underscore after functionname?

File: include/blmc_robots/common_header.hpp [hunk 5]
```diff
@@ -11,10 +11,25 @@
 #ifndef COMMON_HEADER_HPP
 #define COMMON_HEADER_HPP
 
+
```
Comment (mwuethri) [inline include/blmc_robots/common_header.hpp]: this filename seems a bit generic, i think it should be more descriptive.

File: include/blmc_robots/common_header.hpp [hunk 6]
```diff
@@ -141,6 +143,28 @@ void print_vector(std::string v_name, Eigen::Ref<Eigen::VectorXd> v)
...
 }
 
+/**
+ * @brief This small structure is used for reading the calibration parameters
+ * for the calibrations demos.
+ * 
+ * @tparam ROBOT_TYPE 
+ */
+template<class ROBOT_TYPE>
+struct ThreadCalibrationData{
```
Comment (mwuethri) [inline include/blmc_robots/common_header.hpp]: brackets should be on a new line

File: include/blmc_robots/common_header.hpp [hunk 7]
```diff
@@ -141,6 +143,28 @@ void print_vector(std::string v_name, Eigen::Ref<Eigen::VectorXd> v)
...
+  ROBOT_TYPE* robot;
+  Eigen::Vector2d joint_index_to_zero;
+
+  ThreadCalibrationData(ROBOT_TYPE* robot_in){
```
Comment (mwuethri) [inline include/blmc_robots/common_header.hpp]: is it a good idea to pass a plain pointer? the user may pass a pointer to robot_in and then destroy the robot_in-object, and the member of your class will point into the void. i think a shared pointer would be safer.

File: demos/common_demo_header.hpp [hunk 8]
```diff
@@ -0,0 +1,20 @@
...
+
+#include "blmc_robots/common_header.hpp"
+
+namespace blmc_robots
+{
+
+
+
+
+}
```
Comment (mwuethri) [inline demos/common_demo_header.hpp]: for reviewing it would be really useful if you create meaningful commits, which completely implement one small feature. if half is in one and the other half in another commit, it makes reviewing much harder.

File: include/blmc_robots/common_programs_header.hpp [hunk 9]
```diff
@@ -0,0 +1,65 @@
+/**
```
Comment (mwuethri) [inline include/blmc_robots/common_programs_header.hpp]: if you change the filename, it would be good to do it in a separate commit. if you change the filename and change the file content at the same time it makes reviewing very hard.

File: demos/demo_finger_torque_control.cpp [hunk 10]
```diff
@@ -30,23 +30,24 @@ struct Hardware
...
+//             hardware.sliders->get_positions());
 
-        hardware.finger->get_observation(t);
-    }
-    return THREAD_FUNCTION_RETURN_VALUE;
-}
+//         hardware.finger->get_observation(t);
+//     }
+//     return THREAD_FUNCTION_RETURN_VALUE;
+// }
```
Comment (mwuethri) [inline demos/demo_finger_torque_control.cpp]: why did you comment this out?

File: demos/common_demo_header.hpp [hunk 11]
```diff
@@ -0,0 +1,38 @@
...
+ * for the calibrations demos.
+ * 
+ * @tparam ROBOT_TYPE 
+ */
+template<class ROBOT_TYPE>
+struct ThreadCalibrationData{
+  ROBOT_TYPE& robot;
+  Eigen::VectorXd joint_index_to_zero;
+
+  ThreadCalibrationData(ROBOT_TYPE& robot_in):robot(robot_in){
```
Comment (mwuethri) [inline demos/common_demo_header.hpp]: ok, i commented on this on another commit, where it was a pointer. the same concern also holds for  a reference.

File: demos/demo_finger_position_control.cpp [hunk 12]
```diff
@@ -56,7 +56,8 @@ static THREAD_FUNCTION_RETURN_TYPE control_loop(
     }
 }
 
-int main(int argc, char **argv)
+/// TODO: fixme 
+int main(int, char **)
```
Comment (mwuethri) [inline demos/demo_finger_position_control.cpp]: why was this change necessary?

**Comment by MaximilienNaveau** [comment]
Thanks for the review, sorry for the chaotic committing, I will update the code .

File: include/blmc_robots/blmc_joint_module.hpp (hunk 3 above)
Comment (MaximilienNaveau) [reply include/blmc_robots/blmc_joint_module.hpp]: I figured that the go_to function should send 0.0 torques once arrived at its destination it is not like a permanent control here. We just want to do a visual check if the joint went to the desired pose.

File: demos/demo_finger_torque_control.cpp (hunk 10 above)
Comment (MaximilienNaveau) [reply demos/demo_finger_torque_control.cpp]: for warning removal, we need to fix or remove the demo.

File: demos/demo_finger_position_control.cpp (hunk 12 above)
Comment (MaximilienNaveau) [reply demos/demo_finger_position_control.cpp]: Warning Hunt

File: demos/common_demo_header.hpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+ */
+void enable_ctrl_c()
+{
+    // make sure we catch the ctrl+c signal to kill the application properly.
+    struct sigaction sigIntHandler;
+    sigIntHandler.sa_handler = my_handler;
+    sigemptyset(&sigIntHandler.sa_mask);
+    sigIntHandler.sa_flags = 0;
+    sigaction(SIGINT, &sigIntHandler, NULL);
+    StopDemos = false;
```
[the remark this answers is not in the document]
Comment (MaximilienNaveau) [reply demos/common_demo_header.hpp]: I wanna re_initialize the boolean in this method to make sure it is in "false" state.
